{
    "expand": "operations,versionedRepresentations,editmeta,changelog,renderedFields",
    "id": "13443595",
    "self": "https://issues.apache.org/jira/rest/api/2/issue/13443595",
    "key": "ARROW-16491",
    "fields": {
        "fixVersions": [],
        "resolution": null,
        "customfield_12312322": null,
        "customfield_12312323": null,
        "customfield_12312320": null,
        "customfield_12310420": "9223372036854775807",
        "customfield_12312321": null,
        "customfield_12312328": null,
        "customfield_12312329": null,
        "customfield_12312326": null,
        "customfield_12310300": null,
        "customfield_12312327": null,
        "customfield_12312324": null,
        "customfield_12312720": null,
        "customfield_12312325": null,
        "lastViewed": null,
        "priority": {
            "self": "https://issues.apache.org/jira/rest/api/2/priority/4",
            "iconUrl": "https://issues.apache.org/jira/images/icons/priorities/minor.svg",
            "name": "Minor",
            "id": "4"
        },
        "labels": [],
        "customfield_12312333": null,
        "customfield_12312334": null,
        "customfield_12313422": "false",
        "customfield_12310310": "2.0",
        "customfield_12312331": null,
        "customfield_12312332": null,
        "aggregatetimeoriginalestimate": null,
        "timeestimate": null,
        "customfield_12312330": null,
        "versions": [
            {
                "self": "https://issues.apache.org/jira/rest/api/2/version/12350712",
                "id": "12350712",
                "description": "",
                "name": "6.0.1",
                "archived": false,
                "released": true,
                "releaseDate": "2021-11-18"
            },
            {
                "self": "https://issues.apache.org/jira/rest/api/2/version/12350591",
                "id": "12350591",
                "description": "",
                "name": "7.0.0",
                "archived": false,
                "released": true,
                "releaseDate": "2022-02-03"
            }
        ],
        "customfield_12311120": null,
        "customfield_12313826": null,
        "customfield_12312339": null,
        "issuelinks": [
            {
                "id": "12639643",
                "self": "https://issues.apache.org/jira/rest/api/2/issueLink/12639643",
                "type": {
                    "id": "10030",
                    "name": "Reference",
                    "inward": "is related to",
                    "outward": "relates to",
                    "self": "https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"
                },
                "inwardIssue": {
                    "id": "13300561",
                    "key": "ARROW-8567",
                    "self": "https://issues.apache.org/jira/rest/api/2/issue/13300561",
                    "fields": {
                        "summary": "[Python] pa.array() sometimes ignore \"safe=False\"",
                        "status": {
                            "self": "https://issues.apache.org/jira/rest/api/2/status/3",
                            "description": "This issue is being actively worked on at the moment by the assignee.",
                            "iconUrl": "https://issues.apache.org/jira/images/icons/statuses/inprogress.png",
                            "name": "In Progress",
                            "id": "3",
                            "statusCategory": {
                                "self": "https://issues.apache.org/jira/rest/api/2/statuscategory/4",
                                "id": 4,
                                "key": "indeterminate",
                                "colorName": "yellow",
                                "name": "In Progress"
                            }
                        },
                        "priority": {
                            "self": "https://issues.apache.org/jira/rest/api/2/priority/4",
                            "iconUrl": "https://issues.apache.org/jira/images/icons/priorities/minor.svg",
                            "name": "Minor",
                            "id": "4"
                        },
                        "issuetype": {
                            "self": "https://issues.apache.org/jira/rest/api/2/issuetype/1",
                            "id": "1",
                            "description": "A problem which impairs or prevents the functions of the product.",
                            "iconUrl": "https://issues.apache.org/jira/secure/viewavatar?size=xsmall&avatarId=21133&avatarType=issuetype",
                            "name": "Bug",
                            "subtask": false,
                            "avatarId": 21133
                        }
                    }
                }
            }
        ],
        "customfield_12313825": null,
        "assignee": null,
        "customfield_12312337": null,
        "customfield_12313823": null,
        "customfield_12312338": null,
        "customfield_12311920": null,
        "customfield_12313822": null,
        "customfield_12312335": null,
        "customfield_12313821": null,
        "customfield_12312336": null,
        "customfield_12313820": null,
        "status": {
            "self": "https://issues.apache.org/jira/rest/api/2/status/1",
            "description": "The issue is open and ready for the assignee to start work on it.",
            "iconUrl": "https://issues.apache.org/jira/images/icons/statuses/open.png",
            "name": "Open",
            "id": "1",
            "statusCategory": {
                "self": "https://issues.apache.org/jira/rest/api/2/statuscategory/2",
                "id": 2,
                "key": "new",
                "colorName": "blue-gray",
                "name": "To Do"
            }
        },
        "components": [
            {
                "self": "https://issues.apache.org/jira/rest/api/2/component/12328936",
                "id": "12328936",
                "name": "Python"
            }
        ],
        "customfield_12312026": null,
        "customfield_12312023": null,
        "customfield_12312024": null,
        "aggregatetimeestimate": null,
        "customfield_12312022": null,
        "customfield_12310921": null,
        "customfield_12310920": "9223372036854775807",
        "customfield_12312823": null,
        "creator": {
            "self": "https://issues.apache.org/jira/rest/api/2/user?username=blo%40activeviam.com",
            "name": "blo@activeviam.com",
            "key": "JIRAUSER289097",
            "avatarUrls": {
                "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
            },
            "displayName": "LOUSSOUARN Brieuc",
            "active": true,
            "timeZone": "Etc/UTC"
        },
        "subtasks": [],
        "reporter": {
            "self": "https://issues.apache.org/jira/rest/api/2/user?username=blo%40activeviam.com",
            "name": "blo@activeviam.com",
            "key": "JIRAUSER289097",
            "avatarUrls": {
                "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
            },
            "displayName": "LOUSSOUARN Brieuc",
            "active": true,
            "timeZone": "Etc/UTC"
        },
        "aggregateprogress": {
            "progress": 0,
            "total": 0
        },
        "customfield_12313520": null,
        "customfield_12310250": null,
        "progress": {
            "progress": 0,
            "total": 0
        },
        "customfield_12313924": null,
        "votes": {
            "self": "https://issues.apache.org/jira/rest/api/2/issue/ARROW-16491/votes",
            "votes": 0,
            "hasVoted": false
        },
        "worklog": {
            "startAt": 0,
            "maxResults": 20,
            "total": 0,
            "worklogs": []
        },
        "customfield_12313920": null,
        "issuetype": {
            "self": "https://issues.apache.org/jira/rest/api/2/issuetype/1",
            "id": "1",
            "description": "A problem which impairs or prevents the functions of the product.",
            "iconUrl": "https://issues.apache.org/jira/secure/viewavatar?size=xsmall&avatarId=21133&avatarType=issuetype",
            "name": "Bug",
            "subtask": false,
            "avatarId": 21133
        },
        "timespent": null,
        "customfield_12314020": "{summaryBean=com.atlassian.jira.plugin.devstatus.rest.SummaryBean@4d7ff00f[summary={pullrequest=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@6671ec45[overall=PullRequestOverallBean{stateCount=0, state='OPEN', details=PullRequestOverallDetails{openCount=0, mergedCount=0, declinedCount=0}},byInstanceType={}], build=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@4cc8df2[overall=com.atlassian.jira.plugin.devstatus.summary.beans.BuildOverallBean@6db64aea[failedBuildCount=0,successfulBuildCount=0,unknownBuildCount=0,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], review=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@411d0050[overall=com.atlassian.jira.plugin.devstatus.summary.beans.ReviewsOverallBean@7840a802[stateCount=0,state=<null>,dueDate=<null>,overDue=false,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], deployment-environment=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@180309e5[overall=com.atlassian.jira.plugin.devstatus.summary.beans.DeploymentOverallBean@47bdf115[topEnvironments=[],showProjects=false,successfulCount=0,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], repository=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@6e8b6632[overall=com.atlassian.jira.plugin.devstatus.summary.beans.CommitOverallBean@1376d359[count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], branch=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@133aacf3[overall=com.atlassian.jira.plugin.devstatus.summary.beans.BranchOverallBean@3ba7d2d8[count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}]},errors=[],configErrors=[]], devSummaryJson={\"cachedValue\":{\"errors\":[],\"configErrors\":[],\"summary\":{\"pullrequest\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"stateCount\":0,\"state\":\"OPEN\",\"details\":{\"openCount\":0,\"mergedCount\":0,\"declinedCount\":0,\"total\":0},\"open\":true},\"byInstanceType\":{}},\"build\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"failedBuildCount\":0,\"successfulBuildCount\":0,\"unknownBuildCount\":0},\"byInstanceType\":{}},\"review\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"stateCount\":0,\"state\":null,\"dueDate\":null,\"overDue\":false,\"completed\":false},\"byInstanceType\":{}},\"deployment-environment\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"topEnvironments\":[],\"showProjects\":false,\"successfulCount\":0},\"byInstanceType\":{}},\"repository\":{\"overall\":{\"count\":0,\"lastUpdated\":null},\"byInstanceType\":{}},\"branch\":{\"overall\":{\"count\":0,\"lastUpdated\":null},\"byInstanceType\":{}}}},\"isStale\":false}}",
        "customfield_12314141": null,
        "customfield_12314140": null,
        "project": {
            "self": "https://issues.apache.org/jira/rest/api/2/project/12319525",
            "id": "12319525",
            "key": "ARROW",
            "name": "Apache Arrow",
            "projectTypeKey": "software",
            "avatarUrls": {
                "48x48": "https://issues.apache.org/jira/secure/projectavatar?pid=12319525&avatarId=10011",
                "24x24": "https://issues.apache.org/jira/secure/projectavatar?size=small&pid=12319525&avatarId=10011",
                "16x16": "https://issues.apache.org/jira/secure/projectavatar?size=xsmall&pid=12319525&avatarId=10011",
                "32x32": "https://issues.apache.org/jira/secure/projectavatar?size=medium&pid=12319525&avatarId=10011"
            },
            "projectCategory": {
                "self": "https://issues.apache.org/jira/rest/api/2/projectCategory/13960",
                "id": "13960",
                "description": "Apache Arrow",
                "name": "Arrow"
            }
        },
        "aggregatetimespent": null,
        "customfield_12312520": null,
        "customfield_12312521": "Thu May 19 15:38:31 UTC 2022",
        "customfield_12314422": null,
        "customfield_12314421": null,
        "customfield_12314146": null,
        "customfield_12314420": null,
        "customfield_12314145": null,
        "customfield_12314144": null,
        "customfield_12314143": null,
        "resolutiondate": null,
        "workratio": -1,
        "customfield_12312923": null,
        "customfield_12312920": null,
        "customfield_12312921": null,
        "watches": {
            "self": "https://issues.apache.org/jira/rest/api/2/issue/ARROW-16491/watchers",
            "watchCount": 3,
            "isWatching": false
        },
        "created": "2022-05-06T10:13:38.000+0000",
        "updated": "2022-05-19T15:38:31.000+0000",
        "timeoriginalestimate": null,
        "description": "Hello,\r\nsafe option is working well for scalar but not for list. To reproduce :\r\n{code:python}\r\nimport pandas as pd\r\nimport pyarrow as pa\r\nint_dataframe = pd.DataFrame({\"array\": [[1, 2]]})\r\nfloat_dataframe = pd.DataFrame({\"array\": [[1.5, 2.3]]})\r\nint_table = pa.Table.from_pandas(int_dataframe)\r\n{code}\r\n\r\n{code:python}\r\n>>> int_table\r\npyarrow.Table\r\narray: list<item: int64>\r\n  child 0, item: int64\r\n----\r\narray: [[[1,2]]]\r\n{code}\r\n\r\n{code:python}\r\n# this is working instead of throwing a `ArrowInvalid: ... Conversion failed for column array with type`\r\n>>> table = pa.Table.from_pandas(float_dataframe, schema=int_table.schema) \r\n>>> table\r\npyarrow.Table\r\narray: list<item: int64>\r\n  child 0, item: int64\r\n----\r\narray: [[[1,2]]]\r\n{code}\r\n\r\nBehavior for scalar is correct :\r\n{code:python}\r\nint_dataframe = pd.DataFrame({\"array\": [1]})\r\nfloat_dataframe = pd.DataFrame({\"array\": [1.5]})\r\nint_table = pa.Table.from_pandas(int_dataframe)\r\ntable = pa.Table.from_pandas(float_dataframe, schema=int_table.schema) # raise:\r\n---------------------------------------------------------------------------\r\nArrowInvalid                              Traceback (most recent call last)\r\nInput In [6], in <module>\r\n----> 1 table = pa.Table.from_pandas(float_dataframe, schema=int_table.schema)\r\n      2 table\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/table.pxi:1782, in pyarrow.lib.Table.from_pandas()\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/pandas_compat.py:594, in dataframe_to_arrays(df, schema, preserve_index, nthreads, columns, safe)\r\n    589     return (isinstance(arr, np.ndarray) and\r\n    590             arr.flags.contiguous and\r\n    591             issubclass(arr.dtype.type, np.integer))\r\n    593 if nthreads == 1:\r\n--> 594     arrays = [convert_column(c, f)\r\n    595               for c, f in zip(columns_to_convert, convert_fields)]\r\n    596 else:\r\n    597     arrays = []\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/pandas_compat.py:594, in <listcomp>(.0)\r\n    589     return (isinstance(arr, np.ndarray) and\r\n    590             arr.flags.contiguous and\r\n    591             issubclass(arr.dtype.type, np.integer))\r\n    593 if nthreads == 1:\r\n--> 594     arrays = [convert_column(c, f)\r\n    595               for c, f in zip(columns_to_convert, convert_fields)]\r\n    596 else:\r\n    597     arrays = []\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/pandas_compat.py:581, in dataframe_to_arrays.<locals>.convert_column(col, field)\r\n    576 except (pa.ArrowInvalid,\r\n    577         pa.ArrowNotImplementedError,\r\n    578         pa.ArrowTypeError) as e:\r\n    579     e.args += (\"Conversion failed for column {!s} with type {!s}\"\r\n    580                .format(col.name, col.dtype),)\r\n--> 581     raise e\r\n    582 if not field_nullable and result.null_count > 0:\r\n    583     raise ValueError(\"Field {} was non-nullable but pandas column \"\r\n    584                      \"had {} null values\".format(str(field),\r\n    585                                                  result.null_count))\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/pandas_compat.py:575, in dataframe_to_arrays.<locals>.convert_column(col, field)\r\n    572     type_ = field.type\r\n    574 try:\r\n--> 575     result = pa.array(col, type=type_, from_pandas=True, safe=safe)\r\n    576 except (pa.ArrowInvalid,\r\n    577         pa.ArrowNotImplementedError,\r\n    578         pa.ArrowTypeError) as e:\r\n    579     e.args += (\"Conversion failed for column {!s} with type {!s}\"\r\n    580                .format(col.name, col.dtype),)\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/array.pxi:312, in pyarrow.lib.array()\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/array.pxi:83, in pyarrow.lib._ndarray_to_array()\r\n\r\nFile ~/Documents/chouket/.venv/lib/python3.9/site-packages/pyarrow/error.pxi:99, in pyarrow.lib.check_status()\r\n\r\nArrowInvalid: ('Float value 1.5 was truncated converting to int64', 'Conversion failed for column array with type float64')\r\n{code}\r\n\u00a0\r\n\r\n\u00a0",
        "customfield_10010": null,
        "timetracking": {},
        "customfield_12314523": null,
        "customfield_12314127": null,
        "customfield_12314522": null,
        "customfield_12314126": null,
        "customfield_12314521": null,
        "customfield_12314125": null,
        "customfield_12314520": null,
        "customfield_12314124": null,
        "attachment": [
            {
                "self": "https://issues.apache.org/jira/rest/api/2/attachment/13043324",
                "id": "13043324",
                "filename": "image-2022-05-06-13-59-47-720.png",
                "author": {
                    "self": "https://issues.apache.org/jira/rest/api/2/user?username=blo%40activeviam.com",
                    "name": "blo@activeviam.com",
                    "key": "JIRAUSER289097",
                    "avatarUrls": {
                        "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                        "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                        "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                        "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                    },
                    "displayName": "LOUSSOUARN Brieuc",
                    "active": true,
                    "timeZone": "Etc/UTC"
                },
                "created": "2022-05-06T11:59:47.989+0000",
                "size": 71214,
                "mimeType": "image/png",
                "content": "https://issues.apache.org/jira/secure/attachment/13043324/image-2022-05-06-13-59-47-720.png"
            },
            {
                "self": "https://issues.apache.org/jira/rest/api/2/attachment/13043325",
                "id": "13043325",
                "filename": "image-2022-05-06-14-04-37-954.png",
                "author": {
                    "self": "https://issues.apache.org/jira/rest/api/2/user?username=blo%40activeviam.com",
                    "name": "blo@activeviam.com",
                    "key": "JIRAUSER289097",
                    "avatarUrls": {
                        "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                        "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                        "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                        "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                    },
                    "displayName": "LOUSSOUARN Brieuc",
                    "active": true,
                    "timeZone": "Etc/UTC"
                },
                "created": "2022-05-06T12:04:38.295+0000",
                "size": 139258,
                "mimeType": "image/png",
                "content": "https://issues.apache.org/jira/secure/attachment/13043325/image-2022-05-06-14-04-37-954.png"
            }
        ],
        "customfield_12312340": null,
        "customfield_12314123": null,
        "customfield_12312341": null,
        "customfield_12312220": null,
        "customfield_12314122": null,
        "customfield_12314121": null,
        "customfield_12314120": null,
        "customfield_12314129": null,
        "customfield_12314524": null,
        "customfield_12314128": null,
        "summary": "[Python] Table.from_pandas is doing unsafe cast for float array to int array",
        "customfield_12314130": null,
        "customfield_12310291": null,
        "customfield_12310290": null,
        "customfield_12314138": null,
        "customfield_12314137": null,
        "environment": null,
        "customfield_12314136": null,
        "customfield_12314135": null,
        "customfield_12311020": null,
        "customfield_12314134": null,
        "duedate": null,
        "customfield_12314132": null,
        "customfield_12314131": null,
        "comment": {
            "comments": [
                {
                    "self": "https://issues.apache.org/jira/rest/api/2/issue/13443595/comment/17532779",
                    "id": "17532779",
                    "author": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=alenka",
                        "name": "alenka",
                        "key": "alenkaf",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=34058",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=34058",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=34058",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=34058"
                        },
                        "displayName": "Alenka Frim",
                        "active": true,
                        "timeZone": "Europe/Ljubljana"
                    },
                    "body": "I noticed that this works fine with a normal array:\r\n{code:python}\r\n>>> import pandas as pd\r\n>>> import pyarrow as pa\r\n\r\n# See here there are no double brackets\r\n>>> int_dataframe = pd.DataFrame({\"array\": [1, 2]})\r\n>>> float_dataframe = pd.DataFrame({\"array\": [1.5, 2.3]})\r\n\r\n# Error is raised\r\n>>> int_table = pa.Table.from_pandas(int_dataframe)\r\n>>> table = pa.Table.from_pandas(float_dataframe, schema=int_table.schema) \r\nTraceback (most recent call last):\r\n  File \"<stdin>\", line 1, in <module>\r\n  File \"pyarrow/table.pxi\", line 3457, in pyarrow.lib.Table.from_pandas\r\n    arrays, schema, n_rows = dataframe_to_arrays(\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 594, in dataframe_to_arrays\r\n    arrays = [convert_column(c, f)\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 594, in <listcomp>\r\n    arrays = [convert_column(c, f)\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 581, in convert_column\r\n    raise e\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 575, in convert_column\r\n    result = pa.array(col, type=type_, from_pandas=True, safe=safe)\r\n  File \"pyarrow/array.pxi\", line 312, in pyarrow.lib.array\r\n    return _ndarray_to_array(values, mask, type, c_from_pandas, safe,\r\n  File \"pyarrow/array.pxi\", line 83, in pyarrow.lib._ndarray_to_array\r\n    check_status(NdarrayToArrow(pool, values, mask, from_pandas,\r\n  File \"pyarrow/error.pxi\", line 100, in pyarrow.lib.check_status\r\n    raise ArrowInvalid(message)\r\npyarrow.lib.ArrowInvalid: ('Float value 1.5 was truncated converting to int64\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:177  CheckFloatToIntTruncation(batch[0], *out)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/exec.cc:701  kernel_->exec(kernel_ctx_, batch, &out)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/exec.cc:642  ExecuteBatch(batch, listener)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/function.cc:255  executor->Execute(implicitly_cast_args, &listener)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/cast.cc:229  Cast(Datum(value), to_type, options, ctx)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:351  compute::Cast(*MakeArray(tmp_data), out_type, cast_options, &context)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:470  CastBuffer(input_type, *data, length_, null_bitmap_, null_count_, type_, cast_options_, pool_, data)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:281  ConvertData<ArrowType>(&data)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:856  converter.Convert()', 'Conversion failed for column array with type float64')\r\n{code}\r\nso I think the reason for the behaviour you see is due to the fact that in your example converted table columns are ListArrays.\r\n\r\nBut I do get a different error in this next case:\r\n{code:python}\r\n>>> int_dataframe = pd.DataFrame({\"array\": [[1]]})\r\n>>> float_dataframe = pd.DataFrame({\"array\": [[1.5]]})\r\n>>> table = pa.Table.from_pandas(float_dataframe, schema=int_table.schema)\r\nTraceback (most recent call last):\r\n  File \"<stdin>\", line 1, in <module>\r\n  File \"pyarrow/table.pxi\", line 3457, in pyarrow.lib.Table.from_pandas\r\n    arrays, schema, n_rows = dataframe_to_arrays(\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 594, in dataframe_to_arrays\r\n    arrays = [convert_column(c, f)\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 594, in <listcomp>\r\n    arrays = [convert_column(c, f)\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 581, in convert_column\r\n    raise e\r\n  File \"/Users/alenkafrim/repos/arrow/python/pyarrow/pandas_compat.py\", line 575, in convert_column\r\n    result = pa.array(col, type=type_, from_pandas=True, safe=safe)\r\n  File \"pyarrow/array.pxi\", line 312, in pyarrow.lib.array\r\n    return _ndarray_to_array(values, mask, type, c_from_pandas, safe,\r\n  File \"pyarrow/array.pxi\", line 83, in pyarrow.lib._ndarray_to_array\r\n    check_status(NdarrayToArrow(pool, values, mask, from_pandas,\r\n  File \"pyarrow/error.pxi\", line 100, in pyarrow.lib.check_status\r\n    raise ArrowInvalid(message)\r\npyarrow.lib.ArrowInvalid: ('Could not convert [1.5] with type list: tried to convert to int64\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/python_to_arrow.cc:623  PyValue::Convert(this->primitive_type_, this->options_, value)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/iterators.h:57  func(objects[i], i, &keep_going)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/python_to_arrow.cc:1182  converter->Extend(seq, size)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:325  ConvertPySequence(reinterpret_cast<PyObject*>(arr_), reinterpret_cast<PyObject*>(mask_), py_options, pool_)\\n/Users/alenkafrim/repos/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:856  converter.Convert()', 'Conversion failed for column array with type object')\r\n{code}\r\n\u00a0",
                    "updateAuthor": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=alenka",
                        "name": "alenka",
                        "key": "alenkaf",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=34058",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=34058",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=34058",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=34058"
                        },
                        "displayName": "Alenka Frim",
                        "active": true,
                        "timeZone": "Europe/Ljubljana"
                    },
                    "created": "2022-05-06T11:00:26.179+0000",
                    "updated": "2022-05-06T11:00:26.179+0000"
                },
                {
                    "self": "https://issues.apache.org/jira/rest/api/2/issue/13443595/comment/17532823",
                    "id": "17532823",
                    "author": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=blo%40activeviam.com",
                        "name": "blo@activeviam.com",
                        "key": "JIRAUSER289097",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                        },
                        "displayName": "LOUSSOUARN Brieuc",
                        "active": true,
                        "timeZone": "Etc/UTC"
                    },
                    "body": "Hello,\r\nthe first example is indeed working but it is not what interested my case. You are creating a dataframe with a column of type `int` and my ticket is about column of type `list[int]`:\r\n!image-2022-05-06-13-59-47-720.png!\r\n\r\nto clarify I don't get why on a table with `list<item: int64>`column type in the schema can load `list<item: double>`and casting this list to an int list with `safe=True` without throwing.\r\n!image-2022-05-06-14-04-37-954.png!\r\n\r\n\r\n\r\n\u00a0",
                    "updateAuthor": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=blo%40activeviam.com",
                        "name": "blo@activeviam.com",
                        "key": "JIRAUSER289097",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                        },
                        "displayName": "LOUSSOUARN Brieuc",
                        "active": true,
                        "timeZone": "Etc/UTC"
                    },
                    "created": "2022-05-06T12:04:41.804+0000",
                    "updated": "2022-05-06T12:04:41.804+0000"
                },
                {
                    "self": "https://issues.apache.org/jira/rest/api/2/issue/13443595/comment/17533803",
                    "id": "17533803",
                    "author": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=jorisvandenbossche",
                        "name": "jorisvandenbossche",
                        "key": "jorisvandenbossche",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                        },
                        "displayName": "Joris Van den Bossche",
                        "active": true,
                        "timeZone": "Europe/Brussels"
                    },
                    "body": "It is indeed not doing a safe conversion by default in the nested array case.\r\n\r\nSimplifying the example a bit to focus on just the array conversion (which is done column by column for the pandas.DataFrame -> Table conversion):\r\n\r\n{code:python}\r\n>>> pa.array(np.array([[1.5], [2.5, 3.5]], dtype=object), type=pa.list_(pa.int64()), safe=True)\r\n<pyarrow.lib.ListArray object at 0x7f004fc74700>\r\n[\r\n  [\r\n    1\r\n  ],\r\n  [\r\n    2,\r\n    3\r\n  ]\r\n]\r\n{code}\r\n\r\nI noticed that for the non-nested (primitive) array case, this also depends on whether the input is already an array or a generic list like:\r\n\r\n{code:python}\r\n>>> pa.array(np.array([1.5, 2.5]), type=pa.int64(), safe=True)\r\n...\r\nArrowInvalid: Float value 1.5 was truncated converting to int64\r\n{code}\r\n\r\nvs\r\n\r\n{code:python}\r\n>>> pa.array([1.5, 2.5], type=pa.int64(), safe=True)\r\n<pyarrow.lib.Int64Array object at 0x7f004fc72c40>\r\n[\r\n  1,\r\n  2\r\n]\r\n{code}\r\n\r\nThose two take a different code path ({{numpy_to_array.cc}} vs {{python_to_arrow.cc}}), so apparently the {{safe}} keyword is not properly handled in the second code path.   \r\n(don't know by heart if the nested array case also takes that second code path, in which case it might be the same issue)  \r\nAnother potentially related issue: ARROW-8567\r\n",
                    "updateAuthor": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=jorisvandenbossche",
                        "name": "jorisvandenbossche",
                        "key": "jorisvandenbossche",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                        },
                        "displayName": "Joris Van den Bossche",
                        "active": true,
                        "timeZone": "Europe/Brussels"
                    },
                    "created": "2022-05-09T13:26:44.694+0000",
                    "updated": "2022-05-09T13:26:44.694+0000"
                },
                {
                    "self": "https://issues.apache.org/jira/rest/api/2/issue/13443595/comment/17539625",
                    "id": "17539625",
                    "author": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=jorisvandenbossche",
                        "name": "jorisvandenbossche",
                        "key": "jorisvandenbossche",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                        },
                        "displayName": "Joris Van den Bossche",
                        "active": true,
                        "timeZone": "Europe/Brussels"
                    },
                    "body": "Another example of this was reported in https://github.com/apache/arrow/issues/12652",
                    "updateAuthor": {
                        "self": "https://issues.apache.org/jira/rest/api/2/user?username=jorisvandenbossche",
                        "name": "jorisvandenbossche",
                        "key": "jorisvandenbossche",
                        "avatarUrls": {
                            "48x48": "https://issues.apache.org/jira/secure/useravatar?avatarId=10452",
                            "24x24": "https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452",
                            "16x16": "https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452",
                            "32x32": "https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"
                        },
                        "displayName": "Joris Van den Bossche",
                        "active": true,
                        "timeZone": "Europe/Brussels"
                    },
                    "created": "2022-05-19T15:38:31.708+0000",
                    "updated": "2022-05-19T15:38:31.708+0000"
                }
            ],
            "maxResults": 4,
            "total": 4,
            "startAt": 0
        },
        "customfield_12311820": "0|z124r4:",
        "customfield_12314139": null
    }
}